### DISCUSSION TITLE
Langium and embedded languages

### DISCUSSION THREAD

**Discussion by ejfasting** [eclipse-langium/langium#1452] [General]
Hello!

I'm looking into langium as a possible solution to create a new language support for vscode and monaco editor. 

In my scenario i need to create a webextension (since it must be useable from monaco) and it also requires some embedded languages which is a mix of plaintext, html, css, javascript and my custom language. 

Defining my custom language looks doable, but im a little unsure what is within the capabilities of langium and what is not when it comes to embedded languages (?)

From reading the documentation and other threads here on the discussion-board it seems like langium doesnt really have anything to do with embedding languages, it is strictly for creating support for a language? 
I was wondering if you could somehow create a 'top level language that also contains all embedded ones', but i might be trying to screw a screw with a hammer? I also need to leverage existing LSPs for the other languages, and i highly doubt that is something langium can help with (?) 

The question is kinda open, as i want input on what langium can be used for to support my scenario, and any input/suggestion are appreciated :) 

//eivind

**Comment by msujew** [comment]
Hey @ejfasting,

> I was wondering if you could somehow create a 'top level language that also contains all embedded ones', but i might be trying to screw a screw with a hammer?

Yes, please don't do that, it will lead down a rabbit hole you don't want to go down. It doesn't really have to do with Langium, but rather the general approach of dealing with disjointed super languages.

You probably have seen [vscode's documentation on embedded languages](https://code.visualstudio.com/api/language-extensions/embedded-languages). I would recommend something similar as they do in the [Request Forwarding section](https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding). The main steps to take would be:

1. Identify the language of each document block. Send the content of each block to the corresponding language servers on document change/open. Use a specifically crafted URI for each block (something like `<original-document-uri>#<block-id>`, which vscode also uses for code cells of notebook documents). This way, the language server can differentiate between different blocks inside of the same document.
2. Whenever a LSP request hits the client, use a middleware to identify which language server the current block belongs to.
3. Transform the request information (i.e. adjust range/position information) and delegate the request to the appropriate language server.
4. Accept the request result and transform it back. Then return the result to vscode.

An alternative to the request information transformation would be whitespace padding. I.e. replace all data before the code block to whitespace and remove all data after the code block. That way, range/position information stays accurate across document blocks.

Basically, it all comes down to writing a lot of language client middleware, this isn't necessarily specific for Langium.

Note that the [language services](https://code.visualstudio.com/api/language-extensions/embedded-languages#language-services) way of doing it could also be adapted to Langium almost 1:1 as it is outlined in the documentation. But it assumes complete control over all languages that you want include (including the top language). Since you want to include HTML, CSS and JS, I don't think this will work well unless someone has those language services available for consumption (which might be possible).

**Reply by ejfasting** [reply]
Hello @,

Thank you for the explanation!

So that would mean that in my case langium is probably not the correct framework to use then..

I agree that request forwarding sounds like the most sound option, at least when talking about webextensions. Unfortunately the documentation fron vscode is lacking a little when describing embedded languages in a webextension, and ive found it a little rough to find information on how you would go about doing it. But my initial research tried to make a request service, but i guess request forwarding will simplify it.

Might be we will add an additional language later on, and might use langium to help create that one. 

Thanks again for the input! 

Edit: 
Actually looks like vscode has multiple services to use: 
https://github.com/microsoft/vscode-css-languageservice

If these are the services you were talking about? 
Im thinking it might be a good option if we were to create a new language with langium, and as you say its 1:1 with language services. 

//eivind

**Reply by msujew** [reply]
> If these are the services you were talking about?

Yes, with that in mind, this could work pretty well. The basic idea would be to build a single language server for your language that internally forwards the requests to the appropriate language services. Building a language service with Langium is pretty straightforward, since you can basically just plug the the exact requests you get from the language client to the respective service in your languages service container.

**Reply by ejfasting** [reply]
So, to make it super-clear in my head, I would not use langium unless i decide to introduce a new language (in this case a new language service to be used along with the other language services for html/css/js)? 

My solution will depend heavily on how 'easy' it will be to use the language services from my lsp server, but then at least I know this is not something langium can simplify for me :) 

Thanks a lot for the answers, it helps me a lot in figuring out the details! :) 

//Eivind

**Reply by msujew** [reply]
> So, to make it super-clear in my head, I would not use langium unless i decide to introduce a new language (in this case a new language service to be used along with the other language services for html/css/js)?

Right, while you could use Langium to reimplement an existing language service, it's probably not really economical. Reusing something like the `vscode-css-languageservice` would be preferable in that case. Langium is built for developers who either want to build a completely new language, or need editing services (likely LSP based) for a language that already exists, but has a subpar editing experience.

> My solution will depend heavily on how 'easy' it will be to use the language services from my lsp server, but then at least I know this is not something langium can simplify for me :)

I haven't worked with the html-related language services libraries yet, but I assume they are straightforward to integrate, given the way they are used in vscode. No promises though ;)

**Reply by ejfasting** [reply]
I'll start off with integrating with vscode-html-languageservice from a webextension and take it from there. 

Thanks again for the clarification! :) 

//eivind

**Reply by msujew** [reply]
Ah, before I forget about it. [Volar](https://volarjs.dev) might be interesting for you.

**Reply by ejfasting** [reply]
Hi,

Yes, actually come across/got recommended that in another thread as well. The plan is to use that, and use langium for defining the new language :) 

Thanks for the input!

//eivind

**Reply by ejfasting** [reply]
Hi again @msujew,

Looking closer at Volar and it looks promising for creating embedded languages with existing languages. 
I now want to take it one step further and use langium to define the new language, and somehow 'use it' from the volar project. 

I'm wondering what the correct approach would be here, and if you by any chance have an example lying around? 
Less is more, if you have an example, as i just want to get the fundamentals up and running (giving some completionItems for example)


//eivind

**Reply by msujew** [reply]
I haven't worked with Volar yet, there was just no need. So I can't really give any advice in that regard.

**Reply by ejfasting** [reply]
Ok, thats a shame, but I understand why as well :) 

Looks to me like I would have to run the volar-LSP and langium-LSP side-by-side and fetch what i need from langium-lsp when i need it. Optimally I would like to import the langium-stuff as i would when using the vscode-html-languageservice, but I cant wrap my head around how that would work today

**Reply by msujew** [reply]
You should be able to run your langium-based language without creating a new language server. You can just use your language module (created from `create<language>Services` and use the contained services to answer the LSP requests that are coming to your volar language server. Kind of similar to how the CLI uses Langium.

**Reply by ejfasting** [reply]
aha, I didnt consider that, I'll give it a shot and update the thread (if anyone else sumbles upon this down the line). 

Thanks again! :)

**Reply by ejfasting** [reply]
Update on this: 
It looks like you are correct, I should be able to use the returned services to compute the completionItems. But, langium requires a LangiumDocument to be passed inn to the getCompletion-method (myservice.lsp.completionProvider.getCompletion). 

What i have to work with is the TextDocument, position and token, but from what i've found you are not actually using anything else then document and position when calculating the completionItems (which is the first step to get working).

I could possibly create a langiumDocument myself, based on the interface, but it contains other properties that are mandatory. I'm unsure if it would matter that i just create dummy-data in the other properties or if i have to create this langiumDocument propertly to not break stuff. 

Any advice?

**Reply by msujew** [reply]
It should be pretty simple to just use the `shared.workspace.LangiumDocumentFactory` to create a document from a text source.

**Reply by ejfasting** [reply]
Fantastic! This looks to be working nicely :) 

For future reference:
```export const create = {
  name: 'my-service',
  create(context): ServicePluginInstance {
    return {
      async provideCompletionItems(document, position, token) {
        var langiumDocument = shared.workspace.LangiumDocumentFactory.fromTextDocument(document);
        let params = { textDocument: document, position: position };
        const result = await Suo.lsp.CompletionProvider?.getCompletion(langiumDocument, params);
        return result;
      },
    };
  },
} satisfies ServicePlugin;
```
